src/genericsite/models.py

```diff
@@ -80,6 +82,136 @@ def For(cls, site: T.Union[Site, int]) -> models.Manager:
         return cls.objects.filter(site_id=site_id)
 
 
+######################################################################################
+# Media Objects
+######################################################################################
+class MediaObject(models.Model):
+    """Abstract base class for stored media files (image, audio, video, attachment).
+
+    Every media object has an `image_file` attribute so it can have a thumbnail of some
+    kind to display in visual contexts. Thumbnails of standard sizes can be accessed
+    via this field. We assume images will use this field as the main `content_field`.
+    Other subclasses should use this to store a thumbnail or cover art, and add a new
+    field to store their content file, and set the `content_field` attribute to the
+    name of the new field where their audio/video/attachment is stored.
+
+    If `mime_type` is not populated, it will be guessed when the instance is saved,
+    based on the file name stored in the `content_field` (using `mimetypes.guess_type`).
+    """
+
+    class Meta:
+        abstract = True
+
+    content_field = "image_file"
+
+    title = models.CharField(_("title"), max_length=255, unique=True)
+    description = models.CharField(
+        verbose_name=_("description"), max_length=255, blank=True
+    )
+    image_file = ThumbnailerImageField(
+        _("image file"),
+        width_field="image_width",
+        height_field="image_height",
+        max_length=255,
+    )
+    image_width = models.IntegerField(_("image_width"), blank=True)
+    image_height = models.IntegerField(_("image_height"), blank=True)
+    site = models.ForeignKey(Site, on_delete=models.CASCADE, default=1)
+
+    tags = TaggableManager(blank=True)
+
+    # Open Graph has `type`, Schema.org calls it `encodingFormat`, both recommend MIME
+    # type as the value
+    mime_type = models.CharField(
+        _("MIME type"), max_length=255, db_index=True, blank=True
+    )
+    copyright_holder = models.CharField(
+        _("copyright holder"), max_length=255, blank=True
+    )
+    custom_copyright_notice = HTMLField(
+        _("custom copyright notice"),
+        blank=True,
+        help_text=_(
+            r"include a pair of curly braces {} where you want the year inserted"
+        ),
+    )
+    uploaded_dt = models.DateTimeField(_("when uploaded"), auto_now_add=True)
+    created_dt = models.DateTimeField(_("when taken"), blank=True, null=True)
+
+    @property
+    def copyright_year(self):
+        if self.created_dt:
+            return self.created_dt.year
+        else:
+            return self.uploaded_dt.year
+
+    @property
+    def copyright_notice(self):
```

Comment:
I agree it's not DRY, but factoring it out to a helper class just to inherit it in both places would just make it harder to trace the code later. I opted to repeat myself rather than confuse myself. ;)

Reply:
Agreed; maybe a comment that if you update in one place, also update the other?